### PR TITLE
TST: run array-api-tests in CI

### PR DESCRIPTION
**Pull request by lucascolley**

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -73,3 +73,52 @@ jobs:
...
+          python -m pip install -r array-api-tests/requirements.txt
+
+      - name: Install marray
+        run: |
+          python -m pip install --no-deps -e .
+
+      - name: Setup test suite
+        run: |
+          cd ${GITHUB_WORKSPACE}/array-api-tests
+          git apply ${GITHUB_WORKSPACE}/ci/xp-tests.patch
```
Comment (lucascolley) [inline .github/workflows/ci.yml:120]: This is a bit hacky. I wonder whether I could create a GitHub Action to automate the patching for a given array-api-tests version.

Comment (mdhaber) [reply .github/workflows/ci.yml:120]: That would be very cool! 
I'll make you a maintainer so CI runs for you. 
Will there be a way to skip tests? There will probably be a few that are hard to get working (e.g. if there are some with dlpack).

Comment (lucascolley) [reply .github/workflows/ci.yml:120]: yep, https://github.com/numpy/numpy/blob/main/tools/ci/array-api-xfails.txt

Comment (mdhaber) [reply .github/workflows/ci.yml:120]: OK, ran CI and made you a collaborator so you can play around with this.

Comment (lucascolley) [reply .github/workflows/ci.yml:120]: thanks!

**Comment by lucascolley** [comment]
alright, modulo a few hiccups in array-api-tests, only 19 failures currently!

```
FAILED array_api_tests/test_array_object.py::test_getitem_masking - TypeError: The array_api_strict namespace does not support the dtype 'object'
FAILED array_api_tests/test_array_object.py::test_setitem_masking - TypeError: The array_api_strict namespace does not support the dtype 'object'
FAILED array_api_tests/test_linalg.py::test_tensordot - ExceptionGroup: Hypothesis found 3 distinct failures. (3 sub-exceptions)
FAILED array_api_tests/test_manipulation_functions.py::test_repeat - TypeError: repeats must be an int or array
FAILED array_api_tests/test_manipulation_functions.py::test_unstack - assert False
FAILED array_api_tests/test_operators_and_elementwise_functions.py::test_sqrt - AssertionError: out=(inf+0j), but should be roughly sqrt(x)=(3968032412.8978596+1643612645.4244113j) [sqrt()]
FAILED array_api_tests/test_set_functions.py::test_unique_all - ExceptionGroup: Hypothesis found 4 distinct failures. (4 sub-exceptions)
FAILED array_api_tests/test_set_functions.py::test_unique_counts - ExceptionGroup: Hypothesis found 4 distinct failures. (4 sub-exceptions)
FAILED array_api_tests/test_set_functions.py::test_unique_inverse - ExceptionGroup: Hypothesis found 4 distinct failures. (4 sub-exceptions)
FAILED array_api_tests/test_set_functions.py::test_unique_values - TypeError: The array_api_strict namespace does not support the dtype 'object'
FAILED array_api_tests/test_signatures.py::test_array_method_signature[__dlpack__] - AssertionError: Argument 'stream' missing from signature
FAILED array_api_tests/test_special_cases.py::test_empty_arrays[mean] - TypeError: Only floating-point dtypes are allowed in __truediv__
FAILED array_api_tests/test_special_cases.py::test_empty_arrays[std] - TypeError: Only floating-point dtypes are allowed in __truediv__
FAILED array_api_tests/test_special_cases.py::test_empty_arrays[var] - TypeError: Only floating-point dtypes are allowed in __truediv__
FAILED array_api_tests/test_special_cases.py::test_nan_propagation[mean] - TypeError: Only floating-point dtypes are allowed in __truediv__
FAILED array_api_tests/test_special_cases.py::test_nan_propagation[std] - TypeError: Only floating-point dtypes are allowed in __truediv__
FAILED array_api_tests/test_special_cases.py::test_nan_propagation[var] - TypeError: Only floating-point dtypes are allowed in __truediv__
FAILED array_api_tests/test_statistical_functions.py::test_cumulative_sum - ValueError: operands could not be broadcast together with remapped shapes [original->remapped]: (0,)  and requested shape (1,)
FAILED array_api_tests/test_statistical_functions.py::test_mean - TypeError: Only floating-point dtypes are allowed in __truediv__
```

**Comment by mdhaber** [comment]
What's the easiest way to run these locally? Such that I can use a debugger?

**Comment by lucascolley** [comment]
> What's the easiest way to run these locally? I'd like to be able to run them in my IDE and use the debugger.

wait for the pixi PR to merge, then I'll set it up so you can do `pixi run xp-tests`

**Comment by lucascolley** [comment]
> What's the easiest way to run these locally? I'd like to be able to run them in my IDE and use the debugger.

Can you try checking out this branch and running:

```
pixi run -e=dev xp-tests
```

For subsequent runs, you can use this to skip setting up array-api-tests:

```
pixi run -e=dev --skip-deps xp-tests
```

**Comment by mdhaber** [comment]
When I try the first command:
```
(marray) C:\Users\matth\Desktop\marray\mybook>pixi run -e=dev xp-tests
Pixi task (clean-xp-tests in dev): rm -rf array-api-tests

Pixi task (clone-xp-tests in dev): git clone https://github.com/data-apis/array-api-tests.git
Cloning into 'array-api-tests'...
remote: Enumerating objects: 6544, done.
remote: Counting objects: 100% (2388/2388), done.
remote: Compressing objects: 100% (224/224), done.
remote: Total 6544 (delta 2287), reused 2167 (delta 2164), pack-reused 4156 (from 2)
Receiving objects: 100% (6544/6544), 1.61 MiB | 7.72 MiB/s, done.
Resolving deltas: 100% (4811/4811), done.

Pixi task (checkout-xp-tests in dev): git reset --hard 606cc4d11fbcb9abdf425ea6bfe21405228d1d89 && git submodule update --init
HEAD is now at 606cc4d Merge pull request #320 from ev-br/can_cast_cmplx
Submodule 'array_api_tests/array-api' (https://github.com/data-apis/array-api/) registered for path 'array-api'
Cloning into 'C:/Users/matth/Desktop/marray/array-api-tests/array-api'...
Submodule path 'array-api': checked out '25e717735b72c6b55080f48af634015d94833838'

Pixi task (patch-xp-tests in dev): git apply ../ci/xp-tests.patch
error: corrupt patch at line 6
```

**Comment by lucascolley** [comment]
I think one of the pre-commit hooks may have changed some of the characters in the patch for some reason and corrupted it. Let me recreate it.

**Comment by lucascolley** [comment]
@keewis do you know how we can tell the `trailing-whitespace` hook to ignore the whitespace displayed in https://github.com/mdhaber/marray/pull/47/commits/7ed7822e195d630c759f275617bc1579e8ff89d2 ? It seems to be necessary for Windows to think that the patch isn't corrupted.

**Comment by lucascolley** [comment]
@mdhaber please try again now that I've fixed the patch!

**Comment by mdhaber** [comment]
OK, this seems to work in console, and even better, I seem to be able to run the tests and get the failures in my IDE.

**Comment by mdhaber** [comment]
Regarding the failures, I will see if there are any easy ones to fix in this PR. I think we will want to xfail the `unique` ones since `unique` support is pending decisions in gh-28.

**Comment by mdhaber** [comment]
One thing that I'm not sure about the *correct* way to fix is imports from `array_api_test` in the test modules themselves. For instance, `test_special_cases.py` has `from array_api_tests.typing import Array, DataType`. Should I `pip install -e .` in the `array_api_tests` directory?

**Comment by lucascolley** [comment]
> One thing that I'm not sure about the correct way to fix is imports from array_api_test in the test modules themselves. For instance, test_special_cases.py has from array_api_tests.typing import Array, DataType. Should I pip install -e . in the array_api_tests directory?

That should work. I don't think we could make a pixi environment for that without vendoring array-api-tests in this repo.

**Comment by mdhaber** [comment]
```
> (marray) C:\Users\matth\Desktop\marray\array-api-tests>pip install -e .
ERROR: file:///C:/Users/matth/Desktop/marray/array-api-tests does not appear to be a Python project: neither 'setup.py' nor 'pyproject.toml' found.
```

**Comment by lucascolley** [comment]
> array-api-tests does not appear to be a Python project: neither 'setup.py' nor 'pyproject.toml' found.

I suppose that is a feature request for array-api-tests then. Can you workaround it for now by changing to relative imports?

> I also installed some other required packages like nditer using mamba to get other tests running.)

Which packages did you add? We should add them to the `tests` env (and they should probably also be listed in https://github.com/data-apis/array-api-tests/blob/master/requirements.txt).

**Comment by mdhaber** [comment]
hypothesis, pytest-json-report, and ndindex

**Comment by lucascolley** [comment]
they are already part of the `xp-tests` feature which is included in the `dev` environment - which environment were you in?

**Comment by mdhaber** [comment]
I created my own `marray` mamba environment before there was a proper `environment.yml`, I think. So that one.

**Comment by lucascolley** [comment]
Sorry, I might not be understanding the problem. For `pixi run -e=dev xp-tests`, all of those dependencies should be there. If you want to run other commands in that environment you should first run `pixi shell -e dev`. There shouldn't be any need to use a `mamba` env.

**Comment by mdhaber** [comment]
This is all about running the tests in my IDE, not using pixi from the command line. Was that the missing piece?
My IDE needs to know what environment to be in, and for a long time, it's been my own `marray` environment.

**Comment by lucascolley** [comment]
> My IDE needs to know what environment to be in, and for a long time, it's been my own marray environment.

Right - can you point your IDE to `.pixi/envs/dev/`?

**Comment by mdhaber** [comment]
It doesn't list pixi specifically as a type of environment I can add. Should one of the other options work?

image

Also, I tried adding these to array-api-strict-skips.txt, but I don't think they were skipped:
```
# marray does not support `linalg` extension
array_api_tests/test_linalg.py::test_tensordot

# marray does not correctly define `unique_` functions or `__dlpack__ yet`
array_api_tests/test_set_functions.py::test_unique_all
array_api_tests/test_set_functions.py::test_unique_counts
array_api_tests/test_set_functions.py::test_unique_inverse
array_api_tests/test_set_functions.py::test_unique_values
array_api_tests/test_signatures.py::test_array_method_signature[__dlpack__]
```
Where do these go? (Could you add them, and I'll follow in the future?)

**Comment by lucascolley** [comment]
sounds like you are looking for https://pixi.sh/dev/ide_integration/pycharm/

**Comment by lucascolley** [comment]
> # marray does not support `linalg` extension
> array_api_tests/test_linalg.py::test_tensordot

Tensordot is actually in both the main API standard and the linalg extension: https://data-apis.org/array-api/2023.12/API_specification/linear_algebra_functions.html

**Comment by mdhaber** [comment]
Oops, and yeah, that is in the code now and should work. I'll check it out.

**Comment by lucascolley** [comment]
just 6 failures now!

**Comment by mdhaber** [comment]
Hah yeah well skipping tests will do that : ) But yeah fixing `mean`/`var`/`std` helped.

**Comment by mdhaber** [comment]
Adding a skip for `test_repeat` and merging gh-56 should get all the tests to pass.

**Comment by lucascolley** [comment]
1 failure left for `sqrt`

**Comment by lucascolley** [comment]
```
E                   AssertionError: out=(inf+0j), but should be roughly sqrt(x)=(3968032412.8978596+1643612645.4244113j) [sqrt()]
E                     x=(1.3043818701608976e+19+1.304381650258572e+19j)
E                   assert False
E                    +  where False = isclose_complex((inf+0j), (3968032412.8978596+1643612645.4244113j), 3.4028234663852886e+38)
E                   Falsifying example: test_sqrt(
E                       x=MArray(
E                           Array(1.3043819e+19+1.3043817e+19j, dtype=array_api_strict.complex64),
E                           Array(False, dtype=array_api_strict.bool)
E                       ),
E                   )
```

the reference implementation is `cmath.sqrt`

**Comment by lucascolley** [comment]
hmm, 3.4028234663852886e+38 seems close to the limit for single precision, so maybe NumPy (via array-api-strict) is overflowing where cmath isn't? Anyway, seems unlikely to be an marray bug

**Comment by lucascolley** [comment]
okay, in it goes, nice work!

**Comment by keewis** [comment]
> do you know how we can tell the trailing-whitespace hook to ignore the whitespace displayed in https://github.com/mdhaber/marray/commit/7ed7822e195d630c759f275617bc1579e8ff89d2
Sorry for being super late in responding. It doesn't look important anymore, but I believe you would be able to tell `pre-commit` to exclude `.patch` files for this hook

**Pull request merged**